File: src/main/java/com/siemens/industrialbenchmark/dynamics/goldstone/NLGP.java
/**
Copyright 2016 Siemens AG.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package com.siemens.industrialbenchmark.dynamics.goldstone;

/**
 * Normalized, linearly biased Goldstone Potential
 * @author Alexander Hentschel, Michel Tokic
 *
 */
public class NLGP {

	private static final double u0 = Math.cbrt(1 + Math.sqrt(2)) / Math.sqrt(3);
	private static final double r0 = u0 + 1. / (3.*u0);
	private static final double lmbd = 2. * r0*r0 - Math.pow(r0,4) + 8. * Math.sqrt(2./27.) * r0;
	private static final double norm_alpha = 2. / lmbd;
	private static final double norm_beta = 1. / lmbd;
	private static final double norm_kappa = -8. * Math.sqrt(2./27.) / lmbd;
	private static final double qh_b = -Math.sqrt(1./27.);
	
	/**
	 * Function value of normalized, linearly biased Goldstone Potential
	 * in polar coordinates.
	 * @param r in R
	 * @param phi angle in Radians
	 * @return
	 */
	public double polar_nlgp (final double r, final double phi) {
        final double rsq = r*r; 
		return -norm_alpha * rsq + norm_beta * rsq*rsq + norm_kappa * Math.sin(phi) * r;
	}
	
	/**
	 * returns the radius r0 along phi-axis where NLG has minimal function value, i.e. 
       r0 = argmin_{r} polar_nlgp(r,phi)        
	 * @param phi angle in Radians
	 * @return returns the radius r0 along phi-axis where NLG has minimal function value
	 */
	public double global_minimum_radius(double phi) {
	   final double qh = norm_kappa * Math.abs(Math.sin(phi)) / (8. * norm_beta);
        
       final double r0;

        double signum_phi = Math.signum(Math.sin(phi));
        if(signum_phi==0.0){
        	signum_phi = 1.0;
        }
        
        if (qh < qh_b) {
            final double u2 =  Math.cbrt(-signum_phi*qh + Math.sqrt(qh*qh - 1. / 27.));
            r0 = u2 + 1. / (3.*u2);
        } else {
        	r0 = signum_phi * Math.sqrt(4./3.) * Math.cos(1./3. * Math.acos(-qh*Math.sqrt(27.)) );
        }   
        return r0;      
	}	
}

File: industrial_benchmark_python/goldstone/nlgp.py
# coding=utf-8
from __future__ import division
import numpy as np
from numpy import sqrt, sin, cos, arcsin, arccos, pi
from scipy.special import cbrt
'''
The MIT License (MIT)

Copyright 2017 Siemens AG

Author: Alexander Hentschel

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

class nlgp:

    def __init__(self):
        u0 = cbrt(1 + sqrt(2)) / sqrt(3)
        r0 = u0 + 1 / (3*u0)
        
        lmbd = 2 * r0**2 - r0**4 + 8 * sqrt(2/27.) * r0
        
        self.__norm_alpha = 2 / lmbd
        self.__norm_beta = 1 / lmbd
        self.__norm_kappa = -8 * sqrt(2/27.) / lmbd
        self.__phi_b = pi / 4.
        self.__qh_b = - sqrt(1/27.)
    
    def polar_nlgp(self, r, phi):
        '''
        Function value of normalized, linearly biased Goldstone Potential
        in polar coordinates: 
          * r in R
          * angle in Radians
        '''   
        rsq = np.square(r) 
        return -self.__norm_alpha * rsq   + self.__norm_beta * np.square(rsq) + self.__norm_kappa * sin(phi) * r
    
    def global_minimum_radius(self,phi):
        '''
        returns the radius r0 along phi-axis where NLG has minimal function value, i.e. 
            r0 = argmin_{r} polar_nlgp(r,phi)
        angle phi in Radians
        '''
        f = np.vectorize(self.__global_minimum_radius)
        return f(phi)

    def __global_minimum_radius(self,phi):
        qh = self.__norm_kappa * abs(sin(phi)) / (8 * self.__norm_beta)

        signum_phi = np.sign(sin(phi))

        if signum_phi == 0:
            signum_phi = 1

        if (qh <= self.__qh_b):
            u =  cbrt(-signum_phi*qh + sqrt(qh*qh - 1 / 27))
            r0 = u + 1 / (3*u)
        else:
            r0 = signum_phi * sqrt(4/3) * cos(1/3 * arccos(-qh*sqrt(27)) )
        return r0